### PR TITLE
[WFCORE-3684] Support JDK modules

### PR DESCRIPTION
**Pull request by dmlloyd**

### CODE REVIEW

File: modules/WFCORE-3684_JDK Module Support.adoc [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+* https://issues.jboss.org/browse/MODULES-254[MODULES-254]
+* https://issues.jboss.org/browse/WFBUILD-37[WFBUILD-37]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+* (pending)
```
Comment (mjurc) [inline modules/WFCORE-3684_JDK Module Support.adoc]: Feel free to add me here.

Comment (mjurc) [inline modules/WFCORE-3684_JDK Module Support.adoc]: Feel free to add me here.

File: modules/WFCORE-3684_JDK Module Support.adoc [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+== Requirements
+
+=== Hard Requirements
+
+* Support all standard OpenJDK 9.x platform module names as built-in modules
+* Support ``org.jboss.modules`` as a built-in module
+* Allow the boot module loader to load platform modules like ``java.logging`` or ``jdk.sctp``
+** When running on Java 8, emulate these modules as built-in system classpath aggregations
+** Exclude modules which are not present on Java 11+ such as ``java.transaction`` and ``java.corba``
```
Comment (mjurc) [inline modules/WFCORE-3684_JDK Module Support.adoc]: Eventually adding a list of these modules could be useful (for documentation purposes, if anything).

Comment (dmlloyd) [reply modules/WFCORE-3684_JDK Module Support.adoc]: I've added a list of these modules to the bottom of the doc.

File: modules/WFCORE-3684_JDK Module Support.adoc (hunk 1 above)
Comment (dmlloyd) [reply modules/WFCORE-3684_JDK Module Support.adoc]: Done.

**Comment by rsvoboda** [comment]
out of the curiosity, why are files list1 and list2 included in the PR ?

File: modules/WFCORE-3684_JDK Module Support.adoc [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+=== Other Interested Projects
+
+* WildFly Swarm
+
+== Requirements
+
+=== Hard Requirements
+
+* Support all standard OpenJDK 9.x platform module names as built-in modules
```
Comment (rsvoboda) [inline modules/WFCORE-3684_JDK Module Support.adoc]: Should we really look into Java 9 and 10 with this RFE? Or just focus on Java 8 + Java 11 ea builds?

**Comment by dmlloyd** [comment]
Those two files were included by mistake @rsvoboda; fixed.

File: modules/WFCORE-3684_JDK Module Support.adoc (hunk 3 above)
Comment (dmlloyd) [reply modules/WFCORE-3684_JDK Module Support.adoc]: I don't think it hurts to keep track of all the versions.  Also, it is important for compatibility (not just for users, but for us as well) to support as many of the existing Java 8 APIs (which we don't already replace) as possible, so this means aligning more closely with Java 9 than Java 11 (which is missing some of these APIs and introduces some modules which are not present in 8).

Of course once the container runs on Java 9 or higher, it's the JDK that provides the modules and not us.

File: modules/WFCORE-3684_JDK Module Support.adoc [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+|jdk.scripting.nashorn
+|jdk.sctp
+|jdk.security.auth
+|jdk.security.jgss
+|jdk.unsupported
+|jdk.xml.dom
+|org.jboss.modules
+|=====
+
+.List of modules present in Java 9 and/or 10 but removed in Java 11
```
[the remark this answers is not in the document]
Comment (dmlloyd) [reply modules/WFCORE-3684_JDK Module Support.adoc]: I simply compared the output of ``java --list-modules`` in Java 9 and 11.

File: modules/WFCORE-3684_JDK Module Support.adoc [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+* Allow the boot module loader to load platform modules like ``java.logging`` or ``jdk.sctp``
+** When running on Java 8, emulate these modules as built-in system classpath aggregations
+** Exclude modules which are not present on Java 11+ such as ``java.transaction`` and ``java.corba``
+* Remove the magical ``java`` from the system package list, and rely on standard module dependency resolution instead
+* Always add ``java.base`` as a dependency as per the Java Language Specification
+* Move modules whose names differ from their standard names, such as ``java.corba``
+** Introduce a deprecated compatibility module which has the old name and delegates to the new name
+* Deprecate the existing JDK aggregation modules, such as ``javax.api`` and ``sun.jdk``
+** These modules should be updated to aggregate standard modules e.g. ``java.se`` or ``jdk.*``
+* The ``system`` dependency type will be disallowed in the ``urn:jboss:module:1.8`` schema and later
```
[the remark this answers is not in the document]
Comment (dmlloyd) [reply modules/WFCORE-3684_JDK Module Support.adoc]: It is not necessary.  I'm inclined to do it incrementally on an as-needed basis to minimize friction.

**Comment by jamezp** [comment]
This is merged into WildFly Core https://github.com/wildfly/wildfly-core/pull/3201.

**Pull request merged**